pkg/config/validator.go

```diff
@@ -0,0 +1,158 @@
+package config
+
+import (
+	// blank import for embeds
+	_ "embed"
+	"fmt"
+	"strings"
+
+	"github.com/xeipuuv/gojsonschema"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	defaultVersion = "1.0"
+)
+
+//go:embed data/config_schema_v1.0.json
+var schemaV1 []byte
+
+func getSchema(version string) (gojsonschema.JSONLoader, error) {
+
+	// Default schema
+	currentSchema := schemaV1
+
+	switch version {
+	case defaultVersion:
+		currentSchema = schemaV1
+	}
+
+	return gojsonschema.NewStringLoader(string(currentSchema)), nil
+}
+
+func ValidateConfig(config *Config, version string) error {
+	schemaLoader, err := getSchema(version)
+	if err != nil {
+		return err
+	}
+	dataLoader := gojsonschema.NewGoLoader(config)
+	return ValidateSchema(schemaLoader, dataLoader)
+}
+
+func Validate(yamlConfig string, version string) error {
+	j := []byte(yamlConfig)
+	config, err := yaml.YAMLToJSON(j)
+	if err != nil {
+		return err
+	}
+
+	schemaLoader, err := getSchema(version)
+	if err != nil {
+		return err
+	}
+	dataLoader := gojsonschema.NewStringLoader(string(config))
+	return ValidateSchema(schemaLoader, dataLoader)
+}
+
+func ValidateSchema(schemaLoader, dataLoader gojsonschema.JSONLoader) error {
+	result, err := gojsonschema.Validate(schemaLoader, dataLoader)
+	if err != nil {
+		return err
+	}
+
+	if !result.Valid() {
+		return toError(result)
+	}
+	return nil
+}
+
+func toError(result *gojsonschema.Result) error {
+	err := getMostSpecificError(result.Errors())
+	return err
+}
+
+const (
+	jsonschemaOneOf = "number_one_of"
+	jsonschemaAnyOf = "number_any_of"
+)
+
+func getDescription(err validationError) string {
+	switch err.parent.Type() {
+	case "invalid_type":
+		if expectedType, ok := err.parent.Details()["expected"].(string); ok {
+			return fmt.Sprintf("must be a %s", humanReadableType(expectedType))
+		}
+	case jsonschemaOneOf, jsonschemaAnyOf:
+		if err.child == nil {
+			return err.parent.Description()
+		}
+		return err.child.Description()
+	}
+	return err.parent.Description()
+}
+
+func humanReadableType(definition string) string {
```

Comment:
This stuff is from Docker, isn't it? This should probably be acknowledged at least. It might be a license violation. Maybe there is a way of importing it from Docker instead of copy & pasting.

Reply:
Yes some of this code is from this file: https://github.com/docker/docker-ce/blob/f76280404059080d79fcda620caf8cef5a4a22f7/components/cli/cli/compose/schema/schema.go
The code is available through Apache v2 license: https://github.com/docker/docker-ce/blob/master/LICENSE

I couldn't find a way to import this, if you have any ideas please let me know.

If there is a concern, I can rewrite this code from scratch. Please let me know

Reply:
If you're just using public methods (starting with a capital letter) you should in theory just be able to import the package.

If not, a nod to where the code comes from at the top of the file with a link to the license seems sensible.